### PR TITLE
Improve Cayenne Deletion Vectors with KeyBased support

### PR DESCRIPTION
**Pull request by lukekim**
This pull request introduces support for key-based deletion vectors in addition to the existing position-based deletion strategy. This enables more robust deletion tracking for tables with primary keys, allowing deletions to survive data reorganization. The changes include new data structures, schema definitions, and logic updates to handle both deletion strategies throughout the catalog and deletion vector writing process.

**Deletion Strategy Enhancements:**

* Introduced the `DeletionIdentifier` enum in `deletion.rs` to represent both position-based and key-based row identifiers, along with supporting methods for checking emptiness and length. The `DeletionVectorWriteSpec` and `DeletionVectorWriteResult` structs now use this enum. [[1]](diffhunk://#diff-e7ff7e9a096d453b41071bdef53d03b432f7731fdb3b123d8896779a247ffdf8R45-R115) [[2]](diffhunk://#diff-e7ff7e9a096d453b41071bdef53d03b432f7731fdb3b123d8896779a247ffdf8L74-R125)
* Added logic to handle key-based deletions, including sorting, deduplication, and batch creation. New helper functions `build_position_based_batch` and `build_key_based_batch` construct Arrow batches for each strategy, and separate schema definitions are provided for both. [[1]](diffhunk://#diff-e7ff7e9a096d453b41071bdef53d03b432f7731fdb3b123d8896779a247ffdf8L167-R251) [[2]](diffhunk://#diff-e7ff7e9a096d453b41071bdef53d03b432f7731fdb3b123d8896779a247ffdf8L181-R287) [[3]](diffhunk://#diff-e7ff7e9a096d453b41071bdef53d03b432f7731fdb3b123d8896779a247ffdf8R355-R366) [[4]](diffhunk://#diff-e7ff7e9a096d453b41071bdef53d03b432f7731fdb3b123d8896779a247ffdf8R375-R384)
* Updated the `build_delete_file` function and related catalog logic to record the deletion type and, for position-based deletions, the source data file path. [[1]](diffhunk://#diff-e7ff7e9a096d453b41071bdef53d03b432f7731fdb3b123d8896779a247ffdf8R339) [[2]](diffhunk://#diff-c0638e1008db506deb5511b66778a4a7066ded8ed860661c146fedd36668b0a9L479-R481) [[3]](diffhunk://#diff-c0638e1008db506deb5511b66778a4a7066ded8ed860661c146fedd36668b0a9R491-R494) [[4]](diffhunk://#diff-c0638e1008db506deb5511b66778a4a7066ded8ed860661c146fedd36668b0a9L542-R546) [[5]](diffhunk://#diff-c0638e1008db506deb5511b66778a4a7066ded8ed860661c146fedd36668b0a9R555-R563) [[6]](diffhunk://#diff-c0638e1008db506deb5511b66778a4a7066ded8ed860661c146fedd36668b0a9R918-R924)

**Catalog API and Implementation Updates:**

* Added the `clear_delete_files` method to the `MetadataCatalog` trait and implemented it in `CayenneCatalog`, allowing all delete files for a table to be removed after compaction. [[1]](diffhunk://#diff-04dbbb4da5cbac82e3242c6e6fd2ee45421f92efb2a98af9eafa9bfa8b6e4462R193-R198) [[2]](diffhunk://#diff-c0638e1008db506deb5511b66778a4a7066ded8ed860661c146fedd36668b0a9R573-R585)
* Updated the `DeleteFile` struct and related catalog queries to include `source_data_file_path` and `deletion_type` fields, supporting both deletion strategies. [[1]](diffhunk://#diff-c0638e1008db506deb5511b66778a4a7066ded8ed860661c146fedd36668b0a9L479-R481) [[2]](diffhunk://#diff-c0638e1008db506deb5511b66778a4a7066ded8ed860661c146fedd36668b0a9R491-R494) [[3]](diffhunk://#diff-c0638e1008db506deb5511b66778a4a7066ded8ed860661c146fedd36668b0a9L542-R546) [[4]](diffhunk://#diff-c0638e1008db506deb5511b66778a4a7066ded8ed860661c146fedd36668b0a9R555-R563) [[5]](diffhunk://#diff-c0638e1008db506deb5511b66778a4a7066ded8ed860661c146fedd36668b0a9R918-R924)

**Dependency and Benchmark Additions:**

* Added `arrow-row` as a dependency in both the workspace and the `cayenne` crate, preparing for key-based row encoding. [[1]](diffhunk://#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R86) [[2]](diffhunk://#diff-ed8173495ff8f411b7f4cba15b951eef51094a81f6b272d10bb1ecf9d7730aedR14)
* Added a new benchmark for deletion strategies in `crates/cayenne/Cargo.toml`.

**Testing and Validation:**

* Updated tests to cover both position-based and key-based deletion identifiers, ensuring correct batch writing and reading. [[1]](diffhunk://#diff-e7ff7e9a096d453b41071bdef53d03b432f7731fdb3b123d8896779a247ffdf8L299-R430) [[2]](diffhunk://#diff-e7ff7e9a096d453b41071bdef53d03b432f7731fdb3b123d8896779a247ffdf8L315-R452) [[3]](diffhunk://#diff-c0638e1008db506deb5511b66778a4a7066ded8ed860661c146fedd36668b0a9R695)

These changes lay the groundwork for more flexible and reliable deletion handling in the data catalog, especially for tables with primary keys.

### CODE REVIEW

File: crates/cayenne/tests/keybased_deletion_test.rs [hunk 1]
```diff
@@ -0,0 +1,806 @@
...
+
+    assert_eq!(get_row_count(&ctx, table_name).await?, 2);
+
+    Ok(())
+}
+
+test_with_backends!(test_composite_pk_persistence_after_delete_impl);
+
+// Edge Case 15: String PK - insert after delete (reusing PK)
+// Note: In key-based deletion strategies, deleted PKs remain marked as deleted
```
Comment (phillipleblanc) [inline crates/cayenne/tests/keybased_deletion_test.rs]: This will not work for adding on-conflict support to Cayenne - we need the ability to delete a row and re-add the same row with the same PK and have only the new version show up.

**Review by phillipleblanc — DISMISSED** [review]

**Review by phillipleblanc — APPROVED** [review]
